### PR TITLE
📈Add basic application insight telemetry

### PR DESCRIPTION
**Pull request by yufeih**
#3981

### CODE REVIEW

File: src/docfx/lib/log/Telemetry.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        public static void TrackErrorCount(string code, ErrorLevel level)
+        {
+            s_telemetryClient
+                .GetMetric(new MetricIdentifier(null, $"error", "code", "level", "os", "version", "repo", "branch"))
+                .TrackValue(1, code, level.ToString(), s_os, s_version, s_repo, s_branch);
+        }
+
+        public static void TrackCacheTotal(CacheName name)
```
Comment (OsmondJiang) [inline src/docfx/lib/log/Telemetry.cs]: what's that means?

Comment (OsmondJiang) [reply src/docfx/lib/log/Telemetry.cs]: or TrackCacheHit?

File: src/docfx/restore/Restore.cs [hunk 2]
```diff
@@ -22,6 +22,8 @@ public static async Task Run(string docsetPath, CommandLineOptions options, Repo
             using (Progress.Start("Restore dependencies"))
             {
                 var repository = Repository.Create(docsetPath);
+                Telemetry.SetRepository(repository?.Remote, repository?.Branch);
```
[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/restore/Restore.cs]: this is to help answer questions like what is the average build time for azure? so only root repo is needed.

File: src/docfx/lib/log/Progress.cs [hunk 3]
```diff
@@ -67,9 +67,11 @@ public void Dispose()
             {
                 t_scope.Value = t_scope.Value.Pop(out var scope);
 
-                var elapsedMs = Stopwatch.ElapsedMilliseconds;
+                var elapsed = Stopwatch.Elapsed;
+                var elapsedMs = elapsed.TotalMilliseconds;
                 if (Log.Verbose || elapsedMs > ProgressDelayMs)
                 {
+                    Telemetry.TrackOperationDuration(Name, elapsed);
```
[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/lib/log/Progress.cs]: Right. I think we can first disable this telemetry until we need more details in the future.

File: src/docfx/lib/log/Telemetry.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/log/Telemetry.cs]: Tracking Hit/Miss count is hard then tracking Miss/Total count, see GitHubUser cache as an example, there is no clear if/else switch, but it is easier to find all total accesses and missed accesses.

**Pull request merged**